nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,222 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for pulling instance state from sled-agents.
+
+use super::common::BackgroundTask;
+use futures::{future::BoxFuture, FutureExt};
+use nexus_db_model::{InvSledAgent, SledInstance};
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::DataStore;
+use omicron_common::backoff::{self, BackoffError};
+use omicron_uuid_kinds::GenericUuid;
+use serde_json::json;
+use sled_agent_client::Client as SledAgentClient;
+use std::future::Future;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+
+/// Background task that periodically checks instance states.
+#[derive(Clone)]
+pub(crate) struct InstanceWatcher {
+    datastore: Arc<DataStore>,
+    resolver: internal_dns::resolver::Resolver,
+    max_retries: NonZeroU32,
+}
+
+const MAX_SLED_AGENTS: NonZeroU32 = unsafe {
+    // Safety: last time I checked, 100 was greater than zero.
+    NonZeroU32::new_unchecked(100)
+};
+
+impl InstanceWatcher {
+    pub(crate) fn new(
+        datastore: Arc<DataStore>,
+        resolver: internal_dns::resolver::Resolver,
+        max_retries: NonZeroU32,
+    ) -> Self {
+        Self { datastore, resolver, max_retries }
+    }
+
+    fn check_instance(
+        &self,
+        opctx: &OpContext,
+        client: &SledAgentClient,
+        instance: SledInstance,
+    ) -> impl Future<Output = ()> + Send + 'static {
+        let instance_id = instance.instance_id();
+        let watcher = self.clone();
+        let opctx = opctx.child(
+            std::iter::once((
+                "instance_id".to_string(),
+                instance_id.to_string(),
+            ))
+            .collect(),
+        );
+        let client = client.clone();
+
+        async move {
+            let InstanceWatcher { datastore, resolver, max_retries } = watcher;
+            slog::trace!(opctx.log, "checking on instance...");
+            let backoff = backoff::retry_policy_internal_service();
+            let mut retries = 0;
+            let rsp = backoff::retry_notify(
+                backoff,
+                || async {
```

Comment:
I kinda agree with @andrewjstone 's comments here -- I'm not sure that we should be retrying in this background task.

I don't think that this task -- instance monitoring -- is ultimately responsible for deciding whether or not a sled is alive. It *is*, however, responsible for updating the state of an instance if the sled rebooted. In this case, if we see:

- Sled policy: not expunged
- Sled agent: not responding

... and we wait... later, we should see:

- Sled policy: not expunged
- Sled agent: responding! and I don't know about your instance.

I don't *think* this requires an inner retry loop, nor a "max retries" check.

As an alternative, if we expunge the sled, we'll see:

- Sled policy: expunged
- Sled agent: <we shouldn't talk to it>

We'll need to sort through #4872 to decide what to do here, but at least, we'll know that the sled being out-to-lunch is an explicitly marked operation, not something we need to infer

Reply:
Okay, regarding this and your comments from https://github.com/oxidecomputer/omicron/pull/5611#discussion_r1578505777 and https://github.com/oxidecomputer/omicron/pull/5611#discussion_r1578523951, here's what I'm currently thinking I'm going to do as part of this PR:

- remove all the retry logic
- add a new table of "sled instance check failures" to the database. for now, i'm imagining that this would be shaped sort of like `(sled_id, instance_id, id_of_the_nexus_performing_the_health_check, time_updated, failure_count)`. we might also want to add a notion of failure kinds to this as well.
- if we ask a sled-agent about an instance we believe it has, and the sled-agent responds with an HTTP error that proactively indicates that it does not know about that instance, mark that instance as `Failed` (as I believe this would indicate that the sled has been rebooted and the instance is gone[^1])
- for all other errors (the ones that currently get retried), such as TCP connections refused/timing out, etc, increment the failure count table for that `(sled_id, instance_id, our_nexus_id)` tuple
- _for now_, all we do is increment that error count. eventually, the error count might be exposed to an operator and/or used to drive automated decisions (e.g. if the failure count for a particular instance is > _n_, but the sled-agent has responded successfully for other instances, restart the instance...if the failure count for all instances on a sled is > _n_, power cycle the sled, etc)

What do you think?

[^1]: Potentially, if we wanted to be maximally distrustful of the `sled-agent`'s understanding of the state of the world, we could also try pinging the actual `propolis-server` process for that instance, to see if _it_ still exists. If the `sled-agent` doesn't believe it exists, it _shouldn't_, but unfortunately, as you probably already know, distributed systems...

Reply:
It does seem like emitting both these failures and failures from clients to sled-agent endpoints elsewhere in Nexus as Oximeter metrics is potentially a good idea. The table of failure counts was just sort of a strawman proposal so that we're doing _something_ with the information here...I'm also happy to just log them and punt on doing anything else in this PR!

Reply:
If you would like to spit these out as a timeseries, it might be helpful to look at the [`oximeter` self-stats](https://github.com/oxidecomputer/omicron/blob/main/oximeter/collector/src/self_stats.rs) module, which reports statistics about its own collections. A very similar kind of reporting could be done here, I think.

Reply:
A quick update on this: I'm currently working on emitting these failure counts as Oximeter metrics. I'll probably want @bnaecker's feedback on how I'm doing that once I push it up for review --- I can ping you then!